### DISCUSSION TITLE
Graphics bug: diagonal snake body parts in board rendering

### DISCUSSION THREAD

**Discussion by thomasweitzel** [BattlesnakeOfficial/feedback#242] [Bug Reports]
The snake's body parts are normally laid out in a rectangular fashion for the games displayed on the https://play.battlesnake.com website. But the display seems "messed up" in the ending screen, where body parts are rendered in a diagonal fashion. Doesn't look right.

Game where bug can be seen (ending position): https://play.battlesnake.com/game/d9494138-f8f8-4227-bf53-aae82670c0b5

[Screenshot](https://pureandroid.com/downloads/files/misc/battlesnake-graphics-bug.png)

**Comment by bvanvugt** [comment]
Oh interesting! Thanks for reporting this, I'm working on this code this month.

Here's a gif that shows the issue a bit clearer:
![Kapture 2023-08-10 at 07 15 23](https://github.com/BattlesnakeOfficial/feedback/assets/1531419/886f4a78-6954-4985-a5ac-735a9327617a)

I think what's happening is the body part being run into is removed from the snake body render, so the body 'jumps' over that square unintentionally. This was caused by an update to the board render algorithm last month.

That body segment was removed from the polyline to support historical behaviour of showing the snake's head clearly -- but I wonder if that's still necessary. I don't think it's the end of the world if the snake head is drawn over a body segment in this case... 🤔

**Comment by bvanvugt** [comment]
Here's a proposed solution that simply draws the snake head on top of the body segment. It doesn't look *great* but it's certainly less awkward? Let me know what you think.

One issue with this solution is that it's unclear exactly how the snake ran into itself -- you need the context of the previous frame 🤔 

![Kapture 2023-08-11 at 07 23 47](https://github.com/BattlesnakeOfficial/feedback/assets/1531419/dc59ef72-76c5-4181-987c-7f08416f2abc)

**Reply by thomasweitzel** [reply]
Now that **I** understand the solution (and what it "fixes") that **is** implemented, I definitely like it more than the proposed solution - because looking at the last frame alone makes it unclear, what the situation really is. As you said: "you need the context of the previous frame".

Maybe drawing an outline for/on the head in a complementary color (bit-wise complement for RGB colors works well and is quick) would be too much effort? Or making the lower body part a little bit transparent?

If it were my choice between keeping the current situation or implementing the proposed fix, I would choose the former :wink:

**Comment by bvanvugt** [comment]
Tried a few more ideas, but nothing that I love yet...

### Drop Shadow
<img width="243" alt="Screenshot 2023-08-12 at 9 00 30 AM" src="https://github.com/BattlesnakeOfficial/feedback/assets/1531419/d09f2636-7dfb-4717-a7e4-2949283a68d7">

But this breaks when the snake is a darker color.
<img width="241" alt="Screenshot 2023-08-12 at 9 03 05 AM" src="https://github.com/BattlesnakeOfficial/feedback/assets/1531419/bf8c307b-c5d2-406c-9e85-2db52a8570c7">

We could use a white shadow when the snake is dark, but detecting this threshold is challenging and imprecise
<img width="243" alt="Screenshot 2023-08-12 at 9 23 20 AM" src="https://github.com/BattlesnakeOfficial/feedback/assets/1531419/c00c00df-ed74-4fa7-93d4-2e08e99e06d8">


### Lower Head Fill Brightness
```
brightness(0.9);
```
<img width="243" alt="Screenshot 2023-08-12 at 9 05 12 AM" src="https://github.com/BattlesnakeOfficial/feedback/assets/1531419/5c1a22b7-78f5-4c8e-a210-052c4a36d571">

Same problem with darker snakes though.
<img width="240" alt="Screenshot 2023-08-12 at 9 05 32 AM" src="https://github.com/BattlesnakeOfficial/feedback/assets/1531419/bfb2e502-1a9a-45db-9c27-12acb9f7d530">

### Invert Head Fill
```
invert(0.5);
```

<img width="242" alt="Screenshot 2023-08-12 at 9 04 49 AM" src="https://github.com/BattlesnakeOfficial/feedback/assets/1531419/e103b35d-d668-420d-81f8-779ba14fe129">

<img width="241" alt="Screenshot 2023-08-12 at 9 06 06 AM" src="https://github.com/BattlesnakeOfficial/feedback/assets/1531419/11095851-475e-4b07-87f7-937b5dc81ee3">

<img width="239" alt="Screenshot 2023-08-12 at 9 04 23 AM" src="https://github.com/BattlesnakeOfficial/feedback/assets/1531419/343dc18d-cd12-4fee-9247-deea3ec47985">

This works a bit better but looks weird with mid-gray colors obv.

<img width="243" alt="Screenshot 2023-08-12 at 9 04 36 AM" src="https://github.com/BattlesnakeOfficial/feedback/assets/1531419/116c465f-4891-4572-b28e-e95fd6146f80">

### Invert 50% and Darken 10%

```
invert(0.5) brightness(0.9);
```

This is the only effect I can think of that covers all cases, but none of them look *great*...

<img width="245" alt="Screenshot 2023-08-12 at 9 06 19 AM" src="https://github.com/BattlesnakeOfficial/feedback/assets/1531419/a162f116-bc2b-4d02-9b17-246c6450b043">
<img width="241" alt="Screenshot 2023-08-12 at 9 06 06 AM" src="https://github.com/BattlesnakeOfficial/feedback/assets/1531419/e84cde45-b405-41eb-b435-dacf09e7df84">
<img width="248" alt="Screenshot 2023-08-12 at 9 06 32 AM" src="https://github.com/BattlesnakeOfficial/feedback/assets/1531419/f481bb1a-320e-42fe-8fd0-56db117998b0">
<img width="242" alt="Screenshot 2023-08-12 at 9 06 42 AM" src="https://github.com/BattlesnakeOfficial/feedback/assets/1531419/de80fd7a-4421-42e9-9f69-c6d01197a913">

**Reply by bvanvugt** [reply]
Of all these options, I'm most tempted to ship the drop shadow one since it looks the best, even if it's bad for dark colors. Maybe there's another way to solve that 🤔

**Reply by coreyja** [reply]
I like the drop shadow too! The hiding on dark colors is unfortunate, but agreed that maybe it can be handled 'next'!

**Reply by coreyja** [reply]
> Maybe there's another way to solve that 🤔

Just thinking of other options... Could we lighten the body of the snake? That way there is something to drop shadow against even in the case of a dark color? Probably won't make it _super_ visible but maybe not invisible either lol?

**Reply by bvanvugt** [reply]
Yeah that would be interesting to try! Maybe apply a brighten filter to the snake, and a drop shadow to the head.

You'll notice I'm intentionally avoiding parsing and interpreting `snake.color` -- that feels like a rabbit hole of never ending complexity 😅

**Comment by bvanvugt** [comment]
@thomasweitzel I ended up deploying a variation of the shadow solution that looks like this:
<img width="204" alt="Screenshot 2023-08-13 at 12 18 47 PM" src="https://github.com/BattlesnakeOfficial/feedback/assets/1531419/ecd09cee-566c-4c97-9feb-a59bcd8e6a97">

Relevant code is here if you're curious:
https://github.com/BattlesnakeOfficial/board/blob/main/src/lib/components/SvgSnakeHead.svelte#L46

**This change is now deployed, as of Board 1.2.22** 👍
